File: facebook-gamingservices/src/main/java/com/facebook/gamingservices/model/ContextChooseContent.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.gamingservices.model;

import android.os.Parcel;
import androidx.annotation.Nullable;
import com.facebook.share.model.ShareModel;
import com.facebook.share.model.ShareModelBuilder;
import java.util.Collections;
import java.util.List;

public class ContextChooseContent implements ShareModel {

  private final @Nullable List<String> filters;
  private final @Nullable Integer maxSize;
  private final @Nullable Integer minSize;

  private ContextChooseContent(final Builder builder) {
    this.filters = builder.filters;
    this.maxSize = builder.maxSize;
    this.minSize = builder.minSize;
  }

  ContextChooseContent(final Parcel parcel) {
    this.filters = parcel.createStringArrayList();
    this.maxSize = parcel.readInt();
    this.minSize = parcel.readInt();
  }

  public @Nullable List<String> getFilters() {
    return this.filters != null ? Collections.unmodifiableList(this.filters) : null;
  }

  public @Nullable Integer getMaxSize() {
    return this.maxSize;
  }

  public @Nullable Integer getMinSize() {
    return this.minSize;
  }

  @Override
  public void writeToParcel(final Parcel out, final int flags) {
    out.writeStringList(this.filters);
    out.writeInt(this.maxSize);
    out.writeInt(this.minSize);
  }

  @Override
  public int describeContents() {
    return 0;
  }

  /** Builder class for a concrete instance of ContextChooseContent */
  public static class Builder implements ShareModelBuilder<ContextChooseContent, Builder> {
    private @Nullable List<String> filters;
    private @Nullable Integer maxSize;
    private @Nullable Integer minSize;

    /**
     * Sets the set of filters to apply to the context suggestions.
     *
     * @param filters the set of filter to apply
     * @return the builder
     */
    public Builder setFilters(final @Nullable List<String> filters) {
      this.filters = filters;
      return this;
    }

    /**
     * Sets the maximum number of participants that a suggested context should ideally have.
     *
     * @param maxSize the maximum number of participants
     * @return the builder
     */
    public Builder setMaxSize(final @Nullable Integer maxSize) {
      this.maxSize = maxSize;
      return this;
    }

    /**
     * Sets the minimum number of participants that a suggested context should ideally have.
     *
     * @param minSize the minimum number of participants
     * @return the builder
     */
    public Builder setMinSize(final @Nullable Integer minSize) {
      this.minSize = minSize;
      return this;
    }

    @Override
    public ContextChooseContent build() {
      return new ContextChooseContent(this);
    }

    @Override
    public Builder readFrom(final ContextChooseContent content) {
      if (content == null) {
        return this;
      }
      return this.setFilters(content.getFilters())
          .setMaxSize(content.getMaxSize())
          .setMinSize(content.getMinSize());
    }

    Builder readFrom(final Parcel parcel) {
      return this.readFrom(
          (ContextChooseContent)
              parcel.readParcelable(ContextChooseContent.class.getClassLoader()));
    }
  }
}

File: facebook-gamingservices/src/main/java/com/facebook/gamingservices/model/ContextChooseContent.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.gamingservices.model

import android.os.Parcel
import android.os.Parcelable
import com.facebook.share.model.ShareModel
import com.facebook.share.model.ShareModelBuilder
import java.util.Collections

class ContextChooseContent : ShareModel {
  private val filters: List<String>?
  val maxSize: Int?
  val minSize: Int?

  private constructor(builder: Builder) {
    filters = builder.filters
    maxSize = builder.maxSize
    minSize = builder.minSize
  }

  internal constructor(parcel: Parcel) {
    filters = parcel.createStringArrayList()
    maxSize = parcel.readInt()
    minSize = parcel.readInt()
  }

  fun getFilters(): List<String>? = filters?.let { Collections.unmodifiableList(it) }

  override fun writeToParcel(out: Parcel, flags: Int) {
    out.writeStringList(filters)
    out.writeInt(maxSize ?: 0)
    out.writeInt(minSize ?: 0)
  }

  override fun describeContents(): Int = 0

  /** Builder class for a concrete instance of ContextChooseContent */
  class Builder : ShareModelBuilder<ContextChooseContent, Builder> {
    internal var filters: List<String>? = null
    internal var maxSize: Int? = null
    internal var minSize: Int? = null

    /**
     * Sets the set of filters to apply to the context suggestions.
     *
     * @param filters the set of filter to apply
     * @return the builder
     */
    fun setFilters(filters: List<String>?): Builder {
      this.filters = filters
      return this
    }

    /**
     * Sets the maximum number of participants that a suggested context should ideally have.
     *
     * @param maxSize the maximum number of participants
     * @return the builder
     */
    fun setMaxSize(maxSize: Int?): Builder {
      this.maxSize = maxSize
      return this
    }

    /**
     * Sets the minimum number of participants that a suggested context should ideally have.
     *
     * @param minSize the minimum number of participants
     * @return the builder
     */
    fun setMinSize(minSize: Int?): Builder {
      this.minSize = minSize
      return this
    }

    override fun build(): ContextChooseContent = ContextChooseContent(this)

    override fun readFrom(model: ContextChooseContent?): Builder =
        model?.let { setFilters(it.getFilters()).setMaxSize(it.maxSize).setMinSize(it.minSize) }
            ?: this

    fun readFrom(parcel: Parcel): Builder =
        readFrom(
            parcel.readParcelable<Parcelable>(ContextChooseContent::class.java.classLoader)
                as ContextChooseContent?)
  }

  companion object CREATOR : Parcelable.Creator<ContextChooseContent> {
    override fun createFromParcel(parcel: Parcel): ContextChooseContent {
      return ContextChooseContent(parcel)
    }

    override fun newArray(size: Int): Array<ContextChooseContent?> {
      return arrayOfNulls(size)
    }
  }
}
